---
title: "Fluent genomics with plyranges and GenomicRanges"
author: "Stuart Lee and Michael Lawrence"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: true
vignette: >
  %\VignetteIndexEntry{Common Tasks}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE, 
  warning = FALSE,
  message = FALSE
)

fig_path <- dir(system.file("inst", 
                        "vignettes", 
                        package = "plyrangesWorkshops"))

# -- pacakges
library(GenomicRanges)
library(plyranges)
library(HelloRangesData)
```

## Introduction

This tutorial will show you how to think about and perform common genomic data 
wrangling tasks using the using the [plyranges](https://bioconductor.org/packages/release/bioc/html/plyranges.html) and [GenomicRanges](https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html) 
Bioconductor packages. These packages represent genomic data using the 
_GRanges_ data structure, which provides a powerful abstraction for working
with measurements or annotations along genomic regions. In this tutorial
you will use _GRanges_ object to construct workflows to

- read data from common formats like BED 
- find overlaps between different sets of genomic regions
- compute coverage and related summaries

## Who is this tutorial for?

If you are an R beginner or new to the Bioconductor ecosystem, this tutorial
provides a comprehensive overview of one of the most important data structures
in Bioconductor, _GRanges_. We show you how to think of this data structure
as a special kind of table and perform all kinds of operations on it.

If you are familiar with the tidyverse suite of packages, then the operations
and workflows we will develop will look somewhat familiar to you. However, due 
to the complexity of genomics data, plyranges develops a new grammar based
on the tidyverse to meet the needs of the computational biologist. 

We assume you know what a data.frame is and have some knowledge
of vector types in R.

The examples and and exercises are based on the [HelloRanges package tutorial](http://bioconductor.org/packages/release/bioc/vignettes/HelloRanges/inst/doc/tutorial.pdf).
If you are familiar with bedtools, this is an excellent resource for learning
how similar operations are performed with Bioconductor packages.

## Setup 

You will need R (version 3.5 or greater) with the following packages installed:

```{r packages, eval = FALSE}
# Bioconductor
library(GenomicRanges)
library(plyranges)
library(HelloRangesData)
```

If you do not have these packages you can install them as follows:

```{r packages-install, eval = FALSE}
 # -> Bioconductor packages are installed with this package
install.packages("BiocManager")
# then use the install function from BiocManager 
BiocManager::install(c("GenomicRanges", "plyranges", "HelloRangesData"))
```


## Introducing GRanges

The GRanges is a fundamental data structure used in the Bioconductor ecosystem. 
It is very similar to the base R data.frame but with appropriate
semantics for a genomics experiment: it has fixed columns
for the chromosome, start and end coordinates, and the strand, along
with an arbitrary set of additional columns, consisting of
measurements or metadata specific to the data type or experiment.

```{r GRanges, echo = FALSE, fig.cap="An illustration of a GRanges data object for a single sample from an RNA-seq experiment. The core components of the object include a seqnames column (representing the chromosome), a ranges column which consists of start and end coordinates for a genomic region, and a strand identifier (either positive, negative, or unstranded). Metadata are included as columns to the right of the dotted line as annotations (gene-id) or range level covariates (score).", out.width="\\textwidth"}
knitr::include_graphics(file.path(fig_path, "GRanges.png"))
```


GRanges balances flexibility with formal
constraints, so that it is applicable to virtually any genomic
workflow, while also being semantically rich enough to support
high-level operations on genomic ranges.

We can create a GRanges from an ordinary data.frame with plyranges using
the `as_granges()` function:

```{r}
library(plyranges, quietly = TRUE)
genes <- data.frame(seqnames = "VI", 
                    start = c(3322, 3030, 1437,  5066, 6426, 836),
                    end = c(3846, 3338, 2615, 5521, 7565, 1363),
                    strand = c("-", "-", "-", "+", "+", "+"),
                    gene_id=c("YFL064C", "YFL065C", "YFL066C", 
                              "YFL063W", "YFL062W", "YFL067W"),
                    gc = runif(6),
                    score = rbinom(6, size = 3, p = 0.1),
                    stringsAsFactors = FALSE)
genes <- as_granges(genes)

genes
```

You might notice a few  differences between how the `genes` GRanges
object is printed compared to an ordinary data.frame or tibble.  

The measurements/annotation columns like gene_id, gc and score are placed
on the right hand side of the "|". These are referred to as metadata columns;
they are not strictly required to construct a GRanges object but are useful
for providing biological context.

The required part of a GRanges object are left of the "|" 
character. These are the seqnames column, the ranges (start, end, width) column,
and the strand column. 

You may have also noticed that underneath each column name is the type of
that column. The metadata columnns are of some usual base R types like 
`character`, `numeric`, and `integer` but the core columns may be
unfamiliar. The seqnames and strand column are (factor) `Rle` columns.
`Rle` stands for run length encoding, and refers to how a column is stored.
For now, there is no harm in thinking about these as ordinary factors.

The ranges column is an `IRanges` which really consists 
of three columns: the start, end and width coordinates of the range. 
This becomes apparent if you look back to our original data.frame - the start
and end columns from there are just the start and end intervals of the ranges 
column. 

Do not be concerned if this a bit confusing for now, once we start
playing with our GRanges object, everything will become clearer.


## GRanges transformation and import

Once you have obtained data from an experiment or a public repository,
you often want to compute summaries, create new columns, modify the
genomic regions or just rearrange the data. Here you will learn how 
to achieve these tasks using the plyranges:

* Modify, subset, and aggregate genomic data with the `mutate()`, 
`filter()`, and `summarise()`.
* Modify genomic regions with the `mutate()`...
* And  while fixing the start/end/center coordinates with the `anchor_*` family
of functions.
* Merge all overlapping and adjacent genomic regions with `reduce_ranges()`.
* Merge the end points of all genomic regions with `disjoin_ranges()`.
* Sort genomic ranges with `arrange()`.


Any of the above operations can be performed on subsets of the
data with `group_by()`.

## Getting started

To begin we will test out some of these operations using a 
GRanges object obtained from a BED file
containing annotated exons from RefSeq from the hg19 genome build. 
To load the file as GRanges we can simply use the `read_bed` function from
plyranges.

```{r load-bed}
exons_bed <- system.file("extdata", "exons.bed", package="HelloRangesData")

# our genome build, requires internet connection
# build <- genome_info("hg19")

# if no connection ---
build <- system.file("extdata", "hg19.genome", package="HelloRangesData") %>%
  read.delim(header = FALSE, col.names = c("seqnames", "width")) %>%
  as_granges(start = 1) %>%
  set_genome_info(genome = "hg19",
                  seqlengths = width(.))

exons <- read_bed(exons_bed, 
                  genome_info = build)
exons
```

The `genome_info()` function automatically propagates the reference
annotations from the hg19 reference as a GRanges. This is important for
ensuring the integrity of any operations involving modifications to 
genomic regions, like overlaps or nearest neighbours. 

## Filter ranges with `filter()`

We can `filter()` ranges based on column values being `TRUE`. This
requires passing the ranges, along with one or more logical expressions.

For example, we can restrict exons to those that are on the
the X chromosome:

```{r filter-01}
filter(exons, seqnames == "chrX")
```

Multiple conditions can be composed together within a call to filter by 
separating the expressions with a comma. To restrict exons to those 
that are on the X chromosome *AND* on the positive strand:

```{r filter-02}
filter(exons, seqnames == "chrX", strand == "+")
```

This is equivalent to
```{r filter-03}
filter(exons, seqnames == "chrX" & strand == "+")
```

But not the same as filtering exons to those on the X chromosome or 
positively stranded.

```{r filter-04}
filter(exons, seqnames == "chrX" | strand == "+")
```

For a comprehensive overview of all R's logical operators see [__section
5.2.2 of R for Data Science__](https://r4ds.had.co.nz/transform.html#filter-rows-with-filter).

### Overlap based filtering

We can also restrict ranges by overlaps using  `filter_by_overlaps()`, 
or exclude non overlapping features using `filter_by_non_overlaps()`.

To see how these work it'll be useful to load another GRanges object, 
this time from a BED file containing CpG islands.

```{r cpg}
cpg_bed <- system.file("extdata", "cpg.bed", package = "HelloRangesData")

cpg <- read_bed(cpg_bed, 
                genome_info = build)
cpg
```

To find exons that do not overlap any cpg islands we 
```{r filter-06}
filter_by_non_overlaps(exons, cpg)
```

Or we could find the cpg islands that are overlapped by exons 

```{r filter-07}
filter_by_overlaps(cpg, exons)
```

## Modify metadata with `mutate()`

New metadata columns can be created using the `mutate()` function. The first
argument is the GRanges object we are mutating and the remaining arguments
are name-value pairs, where name refers to name of the new column being 
created and value refers to the value the new column will take.

As an example, we make a new column called total_length which contains the number
of bases of in each CpG island. This is just a copy of the width column in 
Ranges object.
```{r mutate-01}
mutate(cpg, total_length =  width)
```

Multiple columns can be computed inside of a call to `mutate()` by separating
each pair of name-value expressions with a comma, we could add an additional
column called mid, which is the midpoint of the CpG island. 

```{r mutate-02}
mutate(cpg,
       total_length = width,
       mid = start + (width - 1) %/% 2
       )
```

Moreover, `mutate()` keeps track of which columnns have been created, so
you can create new columns using pre-existing ones. 

```{r mutate-03}
mutate(cpg,
       total_length = width,
       mid = start + (width - 1) %/% 2,
       mid2 = sqrt(mid)
       )
```


In plyranges and GenomicRanges there are additional helpers that are useful
for column creation during data analysis. As an example, we can 
add the count of the number of times each cpg island overlaps an exon and
add a logical column asking if a cpg island is contained entirely within an 
exon.

```{r}
cpg2 <- mutate(cpg, 
               n_olap = count_overlaps(cpg, exons),
               is_contained = cpg %within% exons
               )

cpg2
```


## Composing workflows with the pipe `%>%`

Now that you can use `filter()` and `mutate()`, you might like to start 
thinking about how you can combine them to form workflows. One way we can do
this is with the pipe operator, `%>%`. 

The pipe operator is read as the word _then_:

```{r pipe-01}
# take cpg then filter
cpg %>% 
  filter(seqnames == "chrX")
```

The argument on the left hand side of the pipe is forwarded to the first
argument of the function on the right hand side. This can be made explicit
with a `.` 

```{r pipe-02}
cpg %>% 
  filter(., seqnames == "chrX")
```

As long as the the result of applying the pipe returns a GRanges object,
multiple functions can be strung together:

```{r pipe-03}
cpg %>% 
  filter(seqnames == "chrX") %>% 
  filter(strand == "+") %>% 
  mutate(new_col = runif(n())) 
```


The use of the `.` can also make pipelines more succinct, as the argument
on the left hand side can be forwarded to multiple parts of the function
on the right hand side. We can rewrite, the creation of cpg2 above more
compactly as

```{r}
cpg %>%  
  mutate(n_olap = count_overlaps(., exons),
         is_contained = . %within% exons)
```

### Exercises

1. Modify the above example so only CpG islands that are completely
within exons are included in the result.
2. Create a new column in exons called tx_id (the transcript
id is everything before _exon in name).   


## Modifying genomic regions with `mutate()`

We have learnt in the above sections how to add new metadata columns with
`mutate()` but we can also use `mutate() to modify the genomic regions.
Recall that there are three components that make up a genomic region; the
start, end, and the width. 

Let's go back to our original GRanges of yeast genes. What happens
if we add 10 bases to the width? 

```{r anchor-01}
genes

genes %>% 
  mutate(width = width + 10L)
```

All three components are mutually dependent, so in order to preserve the 
integrity of the GRanges structure, either the start or the end positions 
must change. In the case of adding to the width position by default 
the start is fixed and the end is increased. But what if we would like
to modify the width but leave the end of fixed? Or the midpoint fixed? 

The solution is to anchor those positions (i.e. leave them fixed), 
using the `anchor_*()` family of functions:
```{r anchor-02}
# fix the end
genes %>% 
  anchor_end() %>% 
  mutate(width = width + 10L)

# fix the centre
genes %>% 
  anchor_centre() %>% 
  mutate(width = width + 10L)
```

There are additional anchors that account for the strandedness of features,
we can fix the 3' or 5' parts of a GRanges object using `anchor_3p()` 
or `anchor_5p()`, respectively.

The plyranges and GenomicRanges API also have utilities to modify 
genomic regions via standard set operations like taking the complement 
(`complement_ranges()`) which will find gaps, or finding flanking ranges
(see `flank_left()` and friends) or promoters (`GenomicRanges::promoters()`).

### Exercises

1. Create a new GRanges object from CpG islands that stretches the intervals
on either side by their width while leaving the centre of the interval fixed. 
2. Create a new GRanges object from exons that has only non-exonic regions.
3. Create a new GRanges object that represent 2bp canonical splice sites
on either side of exon (hint: you can do this with anchoring and mutate 
or with flank).


## Summarise annotations and measurements with `summarise()`

We can compute summarise columns GRanges object using `summarise()`. Using
`summarise()` is very similiar to `mutate()`, we construct new columns
that are summaries of the columns in the input the GRanges. 

```{r}
exons %>% 
  summarise(
    count = n(),
    median_width = median(width)
  )
```

The main difference is that by summarising our data, we lose the necessary
information required to form a valid GRanges. Instead, `summarise()` returns
a `DataFrame` (see `S4Vectors::DataFrame()` for more information). The 
`DataFrame` is the Bioconductor version of the base R data.frame. It works
mostly the same way as the base R `data.frame` but can also hold data structures
that are common in Bioconductor (like `Rle` mentioned above).

### Exercises

1. What is the average length of CpG islands accross the genome?
2. How many distinct transcript identifiers are there in exons?


## Summarise over the genome with `reduce_ranges()` and `disjoin_ranges()`

To preserve the GRanges data structure while performing summaries, it
is often useful to either merge overlapping ranges (called reducing) or expand
ranges by finding the union of the end points or ranges (called disjoining).

To make this more concrete let's `reduce_ranges()` and `disjoin_ranges()` 
on the `genes` object we defined above. 

```{r reduce-01}
# to make things easier to compare 
# sort by the start and end coordinates
genes %>% 
  arrange(start, end)

# reduce_ranges will merge the 3rd and 4th ranges above
genes %>% 
  reduce_ranges()

# disjoin ranges will split the 3rd and 4th ranges into
# 3 disjoint ranges. 
genes %>% 
  disjoin_ranges()
```

Note that by default, `reduce_ranges()` and `disjoin_ranges()` will discard
any strand information. To preserve them use `reduce_ranges_directed()` and
`disjoin_ranges_directed()`.

The `reduce_ranges()` and `disjoin_ranges()` can perform aggregation too.
For example, we can reduce exons and count the number of exons
that were merged together, and concatenate the names of the exons that were 
merged.

```{r reduce-03}
exons %>% 
  reduce_ranges(
    count = n(),
    name = paste(name, collapse = ",")
  )
```

### Exercises

1. Reduce CpG islands, what is the range that has the maximum number of nearby
features?
2. Confirm your results in 1. using `filter_by_overlaps()`.


## Analyses over groups with `group_by()`

Any of the above operations can be performed over subsets of the
`GRanges` object with `group_by()`, this often useful if you want to
`filter()`/`mutate()`/`summarise()`/`reduce_ranges()`/`disjoin_ranges()` 
over annotation identifiers or chromosomes.

```{r group-by-01}
exons %>% 
  group_by(seqnames) %>% 
  summarise(
    count = n(),
    median_width = median(width)
  )
```

### Exercises

Let's read in another BED file that has functional annotations output by
chromHMM from ENCODE human embryonic stem cell ChIP-seq data. 

```{r hesc}
hesc_bed <- system.file("extdata", "hesc.chromHmm.bed", 
                        package="HelloRangesData")

hesc <- read_bed(hesc_bed, genome_info = build)
hesc
```

The name column is the predicted state of a range from chromHMM. 

Use the hesc GRanges object to answer the following:

1. How many ranges are there for each predicted state?
2. How many base pairs are represented in the genome for each predicted state? 
Which state has the maximum number of bases?


## Finding overlaps with the join operator

A common operation on two GRanges is to find overlaps and to keep
metadata associated with both GRanges objects when an overlap has 
occurred. We refer to this operation as an *overlap join*.
There are three types of overlap join: inner (A), intersect (B), and left (C)
that are shown on the figure below.

```{r olap-fig, out.width="\\textwidth"}
knitr::include_graphics(file.path(fig_path, "olaps.png"))
```

The first type of overlap join is called `join_overlap_inner()`, we could 
use this to find which CpG islands overlap exons
```{r olap-01}
cpg2  <- join_overlap_inner(cpg, exons)
cpg2
```

Often we would are also interested in the amount of overlap, we can use
the second type of overlap join `join_overlap_intersect()` to do this.
We can modify the above example, to find which CpG islands overlap exons
and how much the CpG island ranges overlap the exon ranges.

```{r olap-03}
cpg3 <- join_overlap_intersect(cpg, exons)
cpg3
```

Finally, we sometimes would like to know where overlap occurs but also return
any other ranges that are not overlapping. This is this the final type
of overlap join `join_overlap_left()`.

```{r olap-04}
cpg4 <- join_overlap_left(cpg, exons)
cpg4
```

The type of overlap can be modified, by adding suffixes to the function call
for example, to find overlaps that are within we could use 
`join_overlap_inner_within()`.

### Exercises

1. Create a new GRanges object, that has exons that are completely
within Enhancers elements of hesc. How many exons are there?
2. Use `join_overlap_intersect()` to filter exons if at least 50 per cent
of their bases are overlapped by enhancer elements.
3. Count the number of each enhancer element type that are exonic.
There are several ways of doing this, but see if you can come up with a 
solution using `join_overlap_left()` + `disjoin_ranges()`. 

 
## Summarising coverage with `compute_coverage()`

Often we are interested in counting the number of features over
the genome that overlap each other. In `plyranges` this is
done with the `compute_coverage()` function in combination with other functions
to manipulate the results. This function will always return a new GRanges 
with a column called score which contains the coverage value over a range. 

### Coverage values

By default, `compute_coverage()` returns all coverage scores even if there is
no coverage over a genomic region.
```{r cv-1}
cvg <- exons %>% 
  compute_coverage()

cvg

cvg %>% filter(score > 0)
```

Also note that computing coverage will result in a GRanges that is disjoint,
which you can check with `isDisjoint()` from GenomicRanges. 

### Coverage histogram

We can `summarise()` the coverage scores as a histogram over all chromosomes 
in a genome using `group_by()`.

We count the total number of bases (that is, the width of the range) over
each chromosome over each coverage score. 

```{r cv-hist0}
cvg_hist <- cvg %>%
  group_by(seqnames, score) %>% 
  summarise(count = sum(width))
```

We can also compute the coverage histogram over the entire genome by grouping
over the scores. Note that we coerce the width to a numeric
to avoid integer overflow, when summing.

```{r}
cvg_hist_all <- cvg %>% 
  mutate(total_length = as.numeric(width)) %>% 
  group_by(score) %>% 
  summarise(
    seqnames = "genome", 
    count = sum(total_length)
  )
```

Then the results can be put together using `rbind()` or plotted using
your favourite plotting package. 

### A more complicated example

Here we perform another example where we find exons with no coverage that
overlap cpg islands. To compose a chain first we compute the coverage
then use a filter to restrict to zero coverage ranges. Then we
perform an inner overlap join to find which cpg islands overlap the exonic
regions with zero coverage.

```{r}
overlaps <- exons %>% 
  compute_coverage() %>% 
  filter(score == 0L) %>% 
  join_overlap_inner(.,  cpg) 

overlaps
```

We can also compute the coverage histogram of exons over cpg islands
and then plot results as an ecdf.

```{r}
cvg_over_exons <- exons %>% 
  compute_coverage() %>%
  join_overlap_inner(., cpg) %>%
  mutate(
    total_length = as.numeric(width),
    genome_length = sum(total_length)
  ) %>% 
  group_by(score, genome_length) %>%
  summarise(
    count = sum(total_length),
    fraction = sum(total_length) / unique(genome_length)
  ) 

cvg_over_exons

plot(1 - cumsum(fraction) ~ score, 
     data = cvg_over_exons, 
     xlim = c(0,30), 
     type= "s", 
     xlab = "coverage", 
     ylab = "fraction of bp > coverage")
```


## Learning more

We have touched the surface of what's possible with plyranges and
GenomicRanges. A good starting point to learn more about GenomicRanges
is [here](https://bioconductor.github.io/BiocWorkshops/solving-common-bioinformatic-challenges-using-genomicranges.html) and plyranges is [here](https://bioconductor.github.io/BiocWorkshops/fluent-genomic-data-analysis-with-plyranges.html). 

If you have any questions about the use of these packages, please ask
a question on [support.bioconductor.org](https://support.bioconductor.org/) 
or post an issue on [github](https://github.com/sa-lee/plyranges).


## Solutions to exercises

We need to read in three files from HelloRangesData

- exons.bed (RefSeq hg19 annotated exons)
- cpg.bed (RefSeq hg19 annotated CpG islands)
- hesc.chromHmm.bed (predicted functional elements from chromHMM program)

```{r read-bed, eval = FALSE}
suppressPackageStartupMessages(library(plyranges))
# our genome build
build <- system.file("extdata", "hg19.genome", package="HelloRangesData") %>% 
  read.delim(header = FALSE, col.names = c("seqnames", "width")) %>% 
  as_granges(start = 1) %>% 
  set_genome_info(genome = "hg19", 
                  seqlengths = width(.))

exons <- system.file("extdata", "exons.bed", package="HelloRangesData") %>% 
  read_bed(genome_info = build)

cpg <- system.file("extdata", "cpg.bed", package="HelloRangesData") %>% 
  read_bed(genome_info = build)

hesc <- system.file("extdata", "hesc.chromHmm.bed", package="HelloRangesData") %>% 
  read_bed(genome_info = build)

```

### Filtering and mutating

1. Modify the above example so only CpG islands that are completely
within exons are included in the result.

```{r ex-01, eval = FALSE}
cpg %>% 
  mutate(is_contained = . %within% exons) %>% 
  filter(is_contained)

# or just 
# cpg %>% filter(. %within% exons)
```

2. Create a new column in exons called tx_id (the transcript
id is everything before _exon in name).   

```{r, eval = FALSE}
# we can use the sub function in base R
exons <- exons %>% 
  mutate(tx_id = sub("_exon.*", "", name))
```


### Modifying genomic regions

1. Create a new GRanges object from CpG islands that stretches the intervals
on either side by their width while leaving the centre of the interval fixed. 

```{r stretch, eval=FALSE}
cpg_stretch <- cpg %>% 
  anchor_centre() %>% 
  mutate(width = 2*width)

# alternative is to use `stretch()`
```


2. Create a new GRanges object from exons that has only non-exonic regions.

```{r gaps, eval = FALSE}
no_exon <- exons %>% complement_ranges()
```

3. Create a new GRanges object that represent 2bp canonical splice sites
on either side of exon.

```{r flanks, eval = FALSE}
# you can do this directly with mutate and anchors
left <- exons %>% 
  anchor_start() %>% 
  mutate(start = start - 2L, width = 2L)

right <- exons %>% 
  anchor_end() %>% 
  mutate(end = end + 2L,
         width = 2L)

sites <- bind_ranges(list(left = left, right = right), .id = "side")

# or with flank_* family of functions
identical(exons %>% flank_left(2), left)
identical(exons %>% flank_right(2), right)
```

### Summarising GRanges objects

1. How many ranges are there for each predicted state?
2. How many base pairs are represented in the genome for each predicted state? 
Which state has the maximum number of bases?

```{r summarise, eval = FALSE}
state_summary <- hesc %>% 
  group_by(name) %>% 
  summarise(
    n_ranges = n(), 
    n_bases = sum(width)
  )

state_summary %>% 
  as.data.frame() %>% 
  filter(n_bases == max(n_bases))
```

### Overlaps

1. Create a new GRanges object, that has exons that are completely
within Enhancers elements of hesc. How many exons are there?

```{r, eval = FALSE}
enhancers <- hesc %>% 
  filter(grepl("Enhancer", name))

exon_within <- join_overlap_inner_within(exons, enhancers)

exon_within
```

2. Use `join_overlap_intersect()` to filter exons if at least 50 per cent
of their bases are overlapped by enhancer elements.

```{r, eval = FALSE}
exons %>% 
  mutate(total_length = width) %>% 
  join_overlap_intersect(enhancers) %>% 
  filter(width / total_length > 0.5)
```

3. Count the number of each enhancer element type that are exonic.
There are several ways of doing this, but see if you can come up with a 
solution using `join_overlap_left()` + `disjoin_ranges()`. 

```{r left-join, eval = FALSE}
enhancers %>% 
  join_overlap_left(exons) %>% 
  group_by(name.x) %>% 
  disjoin_ranges(olap_any = all(!is.na(tx_id))) %>% 
  group_by(name.x) %>% 
  summarise(prop = sum(olap_any) / n())
```